tests/ct_pipeline.c

```diff
@@ -0,0 +1,136 @@
+    assert(strcmp(reply->str, "OK") == 0);
+    freeReplyObject(reply);
+
+    redisClusterGetReply(cc, (void *)&reply); // reply for: GET foo
+    assert(reply != NULL);
+    assert(reply->type == REDIS_REPLY_STRING);
+    assert(strcmp(reply->str, "5") == 0);
+    freeReplyObject(reply);
+
+    redisClusterGetReply(cc, (void *)&reply); // reply for: GET bar
+    assert(reply != NULL);
+    assert(reply->type == REDIS_REPLY_STRING);
+    assert(strcmp(reply->str, "five") == 0);
+    freeReplyObject(reply);
+
+    redisClusterFree(cc);
+}
+
+//------------------------------------------------------------------------------
+// Async API
+//------------------------------------------------------------------------------
+
+typedef struct ExpectedResult {
+    int type;
+    char *str;
+    bool disconnect;
+} ExpectedResult;
+
+// Callback for Redis connects and disconnects
+void callbackExpectOk(const redisAsyncContext *ac, int status) {
+    UNUSED(ac);
+    assert(status == REDIS_OK);
+}
+
+// Callback for async commands, verifies the redisReply
+void commandCallback(redisClusterAsyncContext *cc, void *r, void *privdata) {
+    redisReply *reply = (redisReply *)r;
+    ExpectedResult *expect = (ExpectedResult *)privdata;
+    assert(reply != NULL);
+    assert(reply->type == expect->type);
+    assert(strcmp(reply->str, expect->str) == 0);
+
+    if (expect->disconnect) {
+        redisClusterAsyncDisconnect(cc);
+    }
+}
+
+// Test of two pipelines using async API
+// In an asynchronous context, commands are automatically pipelined due to the
+// nature of an event loop. Therefore, unlike the synchronous API, there is only
+// a single way to send commands.
+void test_async_pipeline() {
+    redisClusterAsyncContext *acc = redisClusterAsyncContextInit();
+    assert(acc);
+    redisClusterAsyncSetConnectCallback(acc, callbackExpectOk);
+    redisClusterAsyncSetDisconnectCallback(acc, callbackExpectOk);
+    redisClusterSetOptionAddNodes(acc->cc, CLUSTER_NODE);
+    redisClusterConnect2(acc->cc);
+
+    assert(acc->err == 0);
+
+    struct event_base *base = event_base_new();
+    redisClusterLibeventAttach(acc, base);
+
+    int status;
+    ExpectedResult r1 = {.type = REDIS_REPLY_STATUS, .str = "OK"};
+    status = redisClusterAsyncCommand(acc, commandCallback, &r1, "SET foo 6");
+    assert(status == REDIS_OK);
+
+    ExpectedResult r2 = {.type = REDIS_REPLY_STATUS, .str = "OK"};
+    status = redisClusterAsyncCommand(acc, commandCallback, &r2, "SET bar six");
+    assert(status == REDIS_OK);
+
+    ExpectedResult r3 = {.type = REDIS_REPLY_STRING, .str = "6"};
+    status = redisClusterAsyncCommand(acc, commandCallback, &r3, "GET foo");
+    assert(status == REDIS_OK);
+
+    ExpectedResult r4 = {
+        .type = REDIS_REPLY_STRING, .str = "six", .disconnect = true};
+    status = redisClusterAsyncCommand(acc, commandCallback, &r4, "GET bar");
+    assert(status == REDIS_OK);
+
+    event_base_dispatch(base);
```

Comment:
Does this one block until all the callbacks have been called? If yes, why not do disconnect after this line instead of in the last callback?

Reply:
The `event_base_dispatch` function will kick of the eventloop and will hang here forever if not told otherwise. Before this call we have given the even-system a couple of events to work on. The last event will also trigger a call to `redisClusterAsyncDisconnect` which will cleanup hiredis-cluster, all hiredis clients, and tell libevent to exit the loop, and the test will terminate.

Reply:
OK, thanks for the explanation.